src/validators/arguments_v3.rs

```diff
@@ -0,0 +1,654 @@
+use std::str::FromStr;
+
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PyString, PyTuple};
+
+use ahash::AHashSet;
+use pyo3::IntoPyObjectExt;
+
+use crate::build_tools::py_schema_err;
+use crate::build_tools::{schema_or_config_same, ExtraBehavior};
+use crate::errors::{ErrorTypeDefaults, ValError, ValLineError, ValResult};
+use crate::input::{
+    Arguments, BorrowInput, Input, KeywordArgs, PositionalArgs, ValidatedDict, ValidatedTuple, ValidationMatch,
+};
+use crate::lookup_key::LookupKeyCollection;
+use crate::tools::SchemaDict;
+
+use super::validation_state::ValidationState;
+use super::{build_validator, BuildValidator, CombinedValidator, DefinitionsBuilder, Validator};
+
+#[derive(Debug, PartialEq)]
+enum ParameterMode {
+    PositionalOnly,
+    PositionalOrKeyword,
+    VarArgs,
+    KeywordOnly,
+    VarKwargsUniform,
+    VarKwargsUnpackedTypedDict,
+}
+
+impl FromStr for ParameterMode {
+    type Err = PyErr;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "positional_only" => Ok(Self::PositionalOnly),
+            "positional_or_keyword" => Ok(Self::PositionalOrKeyword),
+            "var_args" => Ok(Self::VarArgs),
+            "keyword_only" => Ok(Self::KeywordOnly),
+            "var_kwargs_uniform" => Ok(Self::VarKwargsUniform),
+            "var_kwargs_unpacked_typed_dict" => Ok(Self::VarKwargsUnpackedTypedDict),
+            s => py_schema_err!("Invalid var_kwargs mode: `{}`", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Parameter {
+    name: String,
+    mode: ParameterMode,
+    lookup_key_collection: LookupKeyCollection,
```

Comment:
btw, if we are looking to reduce allocations I think we should consider seeing if we can deduplicate `name` and `lookup_key_collection`, I imagine frequently they hold the same data, copied.

Not for this PR but maybe an interesting thing to put in an issue / follow-up?

Reply:
Will create a follow-up issue